src/main/java/bio/terra/service/dataset/DatasetService.java

```diff
@@ -439,6 +460,98 @@ public List<BulkLoadHistoryModel> getLoadHistory(
     }
   }
 
+  @FunctionalInterface
+  public interface IamAuthorizedCall {
+    void get() throws IamForbiddenException;
+  }
+
+  /**
+   * Throw if the user cannot access the dataset via SAM permissions.
+   *
+   * @param datasetId snapshot UUID
+   * @param userReq authenticated user
+   * @param iamAuthorizedCall throws if dataset inaccessible via SAM permissions
+   */
+  void verifyDatasetAccessible(
+      UUID datasetId, AuthenticatedUserRequest userReq, IamAuthorizedCall iamAuthorizedCall) {
+    boolean iamAuthorized = false;
+    String error = "";
+    try {
+      iamAuthorizedCall.get();
+      iamAuthorized = true;
+    } catch (Exception iamEx) {
+      logger.warn("Snapshot {} inaccessible via SAM for {}", datasetId, userReq.getEmail(), iamEx);
+      error = iamEx.getMessage();
+    } finally {
+      if (!iamAuthorized) {
+        throw new ForbiddenException("Error accessing snapshot: see errorDetails", List.of(error));
+      }
```

Comment:
The only way that `iamAuthorized` is set to true is on line 481. So you could get rid of `iamAuthorized` and move this `throw` into the `catch` clause. You can also get rid of `error` once that's done.